ledger/internal/eval.go

```diff
@@ -172,74 +173,73 @@ func (x *roundCowBase) lookup(addr basics.Address) (ledgercore.AccountData, erro
 		return accountData, nil
 	}
 
-	accountData, _, err := x.l.LookupWithoutRewards(x.rnd, addr)
+	ad, _, err := x.l.LookupWithoutRewards(x.rnd, addr)
 	if err != nil {
 		return ledgercore.AccountData{}, err
 	}
 
-	ad := ledgercore.ToAccountData(accountData)
 	x.accounts[addr] = ad
 	return ad, err
 }
 
 func (x *roundCowBase) lookupAppParams(addr basics.Address, aidx basics.AppIndex) (basics.AppParams, bool, error) {
-	if result, ok := x.appParams[ledgercore.AccountApp{Address: addr, App: aidx}]; ok {
+	aa := ledgercore.AccountApp{Address: addr, App: aidx}
+	if result, ok := x.appParams[aa]; ok {
 		return result.value, result.exists, nil
 	}
 
-	accountData, _, err := x.l.LookupWithoutRewards(x.rnd, addr)
-	if err != nil {
+	resourceData, err := x.l.LookupResource(x.rnd, addr, basics.CreatableIndex(aidx), basics.AppCreatable)
+	if err != nil || resourceData.AppParams == nil {
```

Comment:
should not we cache params miss?

Reply:
correct - we should. But I think that I have another issue here : 
the `x.l.LookupResource` would get you both the owning and holding at the same time.
we should populate both catches with this call.

Reply:
it even better :)

Reply:
oh, I was wondering about that. related, why is it we don't cache missing accounts? I noticed that roundCowBase.lookup() doesn't cache misses but roundCowBase.getCreator does.

Reply:
idk, maybe because of the fact a single missed account usually is an error in evaluator so other txns in a block would not attempt to lookup it.
This does not hold for resources, there are existence checks for assets/apps and missing is not an error. Well, from the other side, existence check is done before creation, so after it goes into state deltas. So looks like not a big deal even for resources.
Let's merge as is ?

Reply:
I'm ok merging it, but I do think that we should do it before the branch is ready.

Reply:
https://github.com/tsachiherman/go-algorand/pull/52